### ISSUE TITLE
Add cmdlets to configure PowerBI Tenant Settings

### ISSUE THREAD

**Issue by ykuijs** [Microsoft/powerbi-powershell#355]
Currently I cannot find any cmdlets that are able to configure the many Power BI Tenant Settings (https://app.powerbi.com/admin-portal/tenantSettings) available in the Power BI Admin portal. Is it possible that configuring Tenant Settings is added?

**Comment by Ventilios** [comment]
That would be really useful. Or as a starting point, having the ability to extract all tenant settings in a convenient way.

**Comment by Shuwi75** [comment]
It seems this functionality (API) is available in the PBI Assessment MIP, although currently there is an auth issue with this particular code.

**Comment by ykuijs** [comment]
Correct, someone provided me with this code that should be able to retrieve information, but am constantly getting a 401 Unauthorized error:

```PowerShell
$auth = Connect-PowerBIServiceAccount
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
$clusterResponse = Invoke-PowerBIRestMethod -Method Get -Url "https://api.powerbi.com/metadata/cluster" | ConvertFrom-Json
$clusterUri = $clusterResponse.backendUrl

$headers = Get-PowerBIAccessToken
$headers."X-PowerBI-User-Admin" = $true

$uri = $clusterUri + "metadata/tenantsettings/properties"
$data = Invoke-RestMethod -Headers $headers -Uri $uri
```

**Comment by Shuwi75** [comment]
That (401) is to be expected at this moment. PG is aware and working on a solution.

**Comment by johann-hackl-microsoft** [comment]
Are there already any updates or rough timelines? 

A customer wants to keep track of his setting history by exporting them on a regular basis.

I only was able to get the settings by PowerShell, when I've manually used the Authorization Token from the Admin Portal using the Edge Developer Tools and then added the code accordingly:
```PowerShell
$headers["Authorization"] = "Bearer ..."
```

**Comment by Shuwi75** [comment]
Beginning of May 2022 I got this feedback "Yes, its in high priority for next semester since its also a popular ask from other customers"; I will check again and post update here.

**Comment by Rajesh-Pradhan** [comment]
> Are there already any updates or rough timelines?
> 
> A customer wants to keep track of his setting history by exporting them on a regular basis.
> 
> I only was able to get the settings by PowerShell, when I've manually used the Authorization Token from the Admin Portal using the Edge Developer Tools and then added the code accordingly:
> 
> ```powershell
> $headers["Authorization"] = "Bearer ..."
> ```

Hi @johann-hackl-microsoft- I've added the headers as you mentioned but I'm still getting 401. Could you please share the complete script which has worked for you.

$auth = Connect-PowerBIServiceAccount
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
$clusterResponse = Invoke-PowerBIRestMethod -Method Get -Url "https://api.powerbi.com/metadata/cluster" | ConvertFrom-Json
$clusterUri = $clusterResponse.backendUrl

$headers = Get-PowerBIAccessToken
$headers."X-PowerBI-User-Admin" = $true
$headers["Authorization"] = "Bearer ....."

$uri = $clusterUri + "metadata/tenantsettings/properties"
$data = Invoke-RestMethod -Headers $headers -Uri $uri

**Comment by johann-hackl-microsoft** [comment]
Sorry, updated full script - somehow the wrong URL made the way into my original comment.
=> use "metadata/tenantsettings" (without "/properties") instead of "metadata/tenantsettings/properties"
Just confirmed with a colleague the following script:

```PowerShell
Set-StrictMode -Version 3.0
Clear-Host

try {
    $headers = Get-PowerBIAccessToken
} catch {
    Connect-PowerBIServiceAccount
    $headers = Get-PowerBIAccessToken
}

[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
$clusterResponse = Invoke-PowerBIRestMethod -Method Get -Url "https://api.powerbi.com/metadata/cluster" | ConvertFrom-Json
$clusterUri = $clusterResponse.backendUrl

# TODO: retrieve token manually from Browser Developer Mode of corresponding Portal Page
$headers["Authorization"    ] = "Bearer eyJ0eX...."
$headers."X-PowerBI-User-Admin" = $true
$headers["x-powerbi-hostenv"] = "Power BI Web"
$uriPostfixList = @(
    "metadata/tenantsettings"
)

foreach ($uriPostfix in $uriPostfixList) {
    $uri = $clusterUri + $uriPostfix
    Write-Output ("Testing: {0}..." -f $uri)
    try {
        $data = Invoke-RestMethod -Headers $headers -Uri $uri
        $data|ConvertTo-Json -depth 10
    } catch {
        Write-Output "### Error"
        $_
        Write-Output "### Exception"
        $_.Exception
    }
}
```

**Comment by abates100** [comment]
Attempting to run this in order to capture tenant settings and have gotten past the 401.   Now however receiving The remote server returned an error: (403) Forbidden at the line Invoke-RestMethod -Headers $headers -Uri $uri.  Does this api need additional auth or parameters if are accessing via a proxy?

**Comment by johann-hackl-microsoft** [comment]
> Attempting to run this in order to capture tenant settings and have gotten past the 401. Now however receiving The remote server returned an error: (403) Forbidden at the line Invoke-RestMethod -Headers $headers -Uri $uri. Does this api need additional auth or parameters if are accessing via a proxy?

Haven't used a proxy when executing the above script, but Invoke-RestMethod (PowerShell Docs: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/invoke-restmethod?view=powershell-7.3) has some proxy related optional arguments - therefore these are worth a try.

And just to be 100% on the same page:
The request is based on an account who sees the tenant setting within browser.
=> # TODO: retrieve token manually from Browser Developer Mode of corresponding Portal Page
